Sources/SourceKitLSP/Swift/CodeActions/ConvertIUOToProperOptional.swift

```diff
@@ -0,0 +1,59 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import SwiftRefactor
+import SwiftSyntax
+
+/// Convert implicitly unwrapped optionals (IUOs) to proper optionals
+@_spi(Testing)
+public struct ConvertIUOToProperOptional: SyntaxRefactoringProvider {
+  public typealias Input = ImplicitlyUnwrappedOptionalTypeSyntax
+  public typealias Context = Void
+  public typealias Output = OptionalTypeSyntax
+
+  @_spi(Testing)
+  public static func refactor(syntax: Input, in context: Context) -> Output? {
+    OptionalTypeSyntax(
+      leadingTrivia: syntax.leadingTrivia,
+      syntax.unexpectedBeforeWrappedType,
+      wrappedType: syntax.wrappedType,
+      syntax.unexpectedBetweenWrappedTypeAndExclamationMark,
+      questionMark: .postfixQuestionMarkToken(
+        leadingTrivia: syntax.exclamationMark.leadingTrivia,
+        trailingTrivia: syntax.exclamationMark.trailingTrivia
+      )
+    )
+  }
+}
+
+extension ConvertIUOToProperOptional: SyntaxRefactoringCodeActionProvider {
+  static let title: String = "Convert Implicitly Unwrapped Optional to Proper Optional"
+
+  static func nodeToRefactor(in scope: SyntaxCodeActionScope) -> Input? {
+    guard let token = scope.innermostNodeContainingRange else {
+      return nil
+    }
+
+    return if let iuoType = token.as(Input.self) ?? token.parent?.as(Input.self) {
+      iuoType
+    } else if token.is(TokenSyntax.self),
+      let wrappedType = token.parent?.as(TypeSyntax.self),
+      let iuoType = wrappedType.parent?.as(Input.self),
+      iuoType.wrappedType == wrappedType
+    {
+      iuoType
+    } else {
+      nil
+    }
```

Comment:
Is there any reason why you didn’t use `findParentOrSelf(ofType:stoppingIf:)` like the other refactoring actions? 

I think the benefit of that would be that the refactoring applies in more situations. Eg. in your test case, you would also get the code action when placing the cursor in `Int`, which seems reasonable to me.

Something like

```suggestion
    return scope.innermostNodeContainingRange?.findParentOfSelf(
      ofType: ImplicitlyUnwrappedOptionalTypeSyntax.self,
      stoppingIf: { $0.is(CodeBlockSyntax.self) || $0.is(MemberBlockSyntax.self) }
    }
```

Reply:
I didn't use `findParentOrSelf(ofType:stoppingIf:)` like other refactoring actions because I wanted to restrict the scope of this code action to the nearest type syntax code that might make local reasoning easier and the namespace of code actions cleaner?

Say, if this code action is triggered by pointing to a type situated deeply inside a heavily nested IUO, it may confuse me a bit.

Reply:
I think it’s fair to assume that usually types are not nested too deeply and in the past we’ve had more issues with being to restrictive with the scopes in which refactorings are shown than being to widely applicable.